File: step_03_arrays/3.2_medium/IB006_flip/solution.py
"""Problem: Flip
Difficulty: MEDIUM | XP: 25
Source: InterviewBit

Given a binary string A, flip exactly one contiguous substring
(change 0->1 and 1->0) to maximize the number of 1s.
Return [L, R] (1-indexed) of the flipped substring.
If no flip needed (all 1s) or no gain possible, return [].
"""
from typing import List

# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n^2)  |  Space: O(1)
# Try every window, calculate net gain from flipping it
# ============================================================
def brute_force(A: str) -> List[int]:
    n = len(A)
    ones = A.count('1')
    best_gain = 0
    best_l = best_r = -1

    for i in range(n):
        gain = 0
        for j in range(i, n):
            gain += 1 if A[j] == '0' else -1
            if gain > best_gain:
                best_gain = gain
                best_l, best_r = i + 1, j + 1

    return [best_l, best_r] if best_l != -1 else []


# ============================================================
# APPROACH 2: OPTIMAL (Transform + Kadane's)
# Time: O(n)  |  Space: O(n)
# Map '0' -> +1, '1' -> -1; find max subarray sum (Kadane's).
# ============================================================
def optimal(A: str) -> List[int]:
    t = [1 if c == '0' else -1 for c in A]
    max_gain = 0
    cur_sum = 0
    temp_start = 0
    best_l = best_r = -1

    for i, v in enumerate(t):
        cur_sum += v
        if cur_sum > max_gain:
            max_gain = cur_sum
            best_l = temp_start + 1
            best_r = i + 1
        if cur_sum < 0:
            cur_sum = 0
            temp_start = i + 1

    return [best_l, best_r] if best_l != -1 else []


# ============================================================
# APPROACH 3: BEST (Kadane's without auxiliary array)
# Time: O(n)  |  Space: O(1)
# Same idea but computes the transform on-the-fly
# ============================================================
def best(A: str) -> List[int]:
    max_gain = 0
    cur_sum = 0
    temp_start = 0
    best_l = best_r = -1

    for i, c in enumerate(A):
        val = 1 if c == '0' else -1
        cur_sum += val
        if cur_sum > max_gain:
            max_gain = cur_sum
            best_l = temp_start + 1
            best_r = i + 1
        if cur_sum < 0:
            cur_sum = 0
            temp_start = i + 1

    return [best_l, best_r] if best_l != -1 else []


if __name__ == "__main__":
    cases = [
        ("010", [1, 1]),       # flip index 1 (0-indexed)
        ("111", []),            # already all 1s
        ("0000", [1, 4]),
        ("010101", [1, 1]),     # any single 0 works; leftmost first
        ("10001", [2, 4]),
    ]
    print("=== Flip Binary String ===")
    for A, exp in cases:
        b = brute_force(A)
        o = optimal(A)
        bst = best(A)
        ok = "OK" if b == o == bst else f"MISMATCH(BF={b},OPT={o},BEST={bst})"
        print(f"'{A}' => BF={b} OPT={o} BEST={bst} EXP={exp} {ok}")

File: step_03_arrays/3.2_medium/IB006_flip/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(n^2)  |  Space: O(1)
// Try every possible contiguous flip window; count 1s after flip.
// Return [L, R] (1-indexed) that maximises 1s, empty if no 0s.
// ============================================================
class BruteForce {
    public static int[] solve(String A) {
        int n = A.length();
        // Count initial 1s
        int ones = 0;
        for (char c : A.toCharArray()) if (c == '1') ones++;

        int bestL = -1, bestR = -1, bestTotal = ones;

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                // Count gain: flip [i..j], each 0 becomes +1, each 1 becomes -1
                int gain = 0;
                for (int k = i; k <= j; k++) gain += (A.charAt(k) == '0') ? 1 : -1;
                if (ones + gain > bestTotal) {
                    bestTotal = ones + gain;
                    bestL = i + 1; bestR = j + 1;
                }
            }
        }
        return bestL == -1 ? new int[]{} : new int[]{bestL, bestR};
    }
}

// ============================================================
// APPROACH 2: OPTIMAL (Transform + Kadane's)
// Time: O(n)  |  Space: O(n)
// Transform: treat '0' as +1 and '1' as -1.
// Flipping [L,R] gains (sum of transformed values).
// Find max-sum subarray of transformed array using Kadane's.
// If max gain <= 0, no flip needed.
// ============================================================
class Optimal {
    public static int[] solve(String A) {
        int n = A.length();
        int[] t = new int[n];
        for (int i = 0; i < n; i++) t[i] = (A.charAt(i) == '0') ? 1 : -1;

        int maxGain = 0, curSum = 0;
        int start = 0, bestL = -1, bestR = -1, tempStart = 0;

        for (int i = 0; i < n; i++) {
            curSum += t[i];
            if (curSum > maxGain) {
                maxGain = curSum;
                bestL = tempStart + 1;
                bestR = i + 1;
            }
            if (curSum < 0) {
                curSum = 0;
                tempStart = i + 1;
            }
        }
        return bestL == -1 ? new int[]{} : new int[]{bestL, bestR};
    }
}

// ============================================================
// APPROACH 3: BEST (Kadane's in-place, no extra array)
// Time: O(n)  |  Space: O(1)
// Same transform but done on-the-fly without allocating array.
// Tie-breaking: prefer leftmost, smallest window per problem spec.
// ============================================================
class Best {
    public static int[] solve(String A) {
        int n = A.length();
        int maxGain = 0, curSum = 0, tempStart = 0;
        int bestL = -1, bestR = -1;

        for (int i = 0; i < n; i++) {
            int val = (A.charAt(i) == '0') ? 1 : -1;
            curSum += val;
            if (curSum > maxGain) {
                maxGain = curSum;
                bestL = tempStart + 1;
                bestR = i + 1;
            }
            if (curSum < 0) {
                curSum = 0;
                tempStart = i + 1;
            }
        }
        return bestL == -1 ? new int[]{} : new int[]{bestL, bestR};
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Flip Binary String ===");
        String[] tests  = {"010", "111", "0000", "1111", "010101"};
        // Expected flips: [1,1], [], [1,4], [], [1,1] or [2,2] or [3,3] etc.
        for (String t : tests) {
            int[] bf   = BruteForce.solve(t);
            int[] op   = Optimal.solve(t);
            int[] best = Best.solve(t);
            System.out.printf("'%s' => BF=%s  OPT=%s  BEST=%s%n",
                t, Arrays.toString(bf), Arrays.toString(op), Arrays.toString(best));
        }
    }
}
